website/src/components/Chat/ChatConversation.tsx

```diff
@@ -165,23 +170,61 @@ export const ChatConversation = memo(function ChatConversation({ chatId: chatIdP
     [sendVote]
   );
 
-  const entries = useMemo(
-    () =>
-      currentThread.map((message) => (
-        <ChatMessageEntry
-          key={message.id}
-          message={message}
-          onVote={handleOnVote}
-          onRetry={handleOnRetry}
-          isSending={isSending}
-        ></ChatMessageEntry>
-      )),
-    [handleOnVote, currentThread, handleOnRetry, isSending]
+  const handleEditPrompt = useCallback(
+    async ({ chatId, parentId, content }: EditPromptParams) => {
+      if (!content || isSending) {
+        return;
+      }
+
+      setIsSending.on();
+      setReytryingParentId(parentId);
+      const prompter_arg: InferencePostPrompterMessageParams = {
+        chat_id: chatId,
+        content,
+        parent_id: parentId,
+      };
+
+      let prompter_message: InferenceMessage | null = null;
+      const dummyMessage: InferenceMessage = {
+        id: "__dummy__",
+        ...prompter_arg,
+        created_at: new Date().toISOString(),
+        role: "prompter",
+        state: "complete",
+        score: 0,
+        reports: [],
+      };
+
+      try {
+        // push the dummy message first to avoid layout shift
```

Comment:
what layout shift happens here? and how does this solve it?

Reply:
To update a tree, we now relying on `currentRetryingParentId`, we travel the tree until we meet`currentRetryingParentId` and replace all of its children. Without the dummy message, everything still work fine. But from the UX perspective, user expect their edited message to be displayed instantly. Without the dummy, the user has to wait a bit, and then we will render the edited message when it available, so the layout will be jumping around for a bit.

Reply:
I see, I wonder if it is necessary to remove the old message in the first place? when we get the response we can remove the last 2 messages and add the new 2? does our architecture allow this?